pkg/dmap-funcs/rate-limiter/global-token-bucket/global-token-bucket.go

```diff
@@ -117,27 +148,48 @@ func (gtb *GlobalTokenBucket) Close() error {
 	if err != nil {
 		return err
 	}
+	err = gtb.jobGroup.DeregisterJob(gtb.name)
+	if err != nil {
+		return err
+	}
 	return nil
 }
 
 func (gtb *GlobalTokenBucket) executeTakeRequest(ctx context.Context, label string, n float64, canWait bool, deadline time.Time) (bool, time.Duration, float64, float64) {
+	getWaitTime := func(availableAt time.Time) time.Duration {
+		var waitTime time.Duration
+		if !availableAt.IsZero() {
+			waitTime = time.Until(availableAt)
+			if waitTime < 0 {
+				waitTime = 0
+			}
+		}
+		return waitTime
+	}
+
 	if gtb.GetPassThrough() {
 		return true, 0, 0, 0
 	}
 
 	c := &counter{}
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	existing, loaded := gtb.counters.LoadOrStore(label, c)
-	if loaded {
-		c = existing.(*counter)
-	}
 
-	if !c.availableAt.IsZero() && time.Now().Before(c.availableAt) {
-		if n >= c.tokens {
-			return false, 0, 0, 0
-		} else {
-			c.tokens -= n
+	if !canWait {
+		existing, loaded := gtb.counters.LoadOrStore(label, c)
+		if loaded {
+			c = existing.(*counter)
+		}
+
+		c.lock.Lock()
+		defer c.lock.Unlock()
+
+		waitTime := getWaitTime(c.availableAt)
+		if c.lastRequestedTokens >= n && waitTime > 0 {
+			remaining := c.remaining
+			current := c.current
+			if n < 0 {
+				gtb.counters.Delete(label)
```

Comment:
Not here. Delete after update

Reply:
[6f6a541](https://github.com/fluxninja/aperture/pull/3155/commits/6f6a5411e982ccef57b3d24df36cc28dcd3cd4ea) addresses this.